rfc-091-sharing-assets.md

```diff
@@ -0,0 +1,43 @@
+# Sharing assets
+
+## Summary
+
+We need to share assets to make sure users only have to download a thing once. To do this, we'll upload each application's assets to the same location.
+
+## Problem
+
+In [RFC84 we proposed replacing Static with a gem](https://github.com/alphagov/govuk-rfcs/pull/84). One of the architectural challenges with this approach is that by moving assets (like CSS, JS and images) into the gem, the assets won't be shared for users. This means that users visiting GOV.UK would have to download assets multiple times, depending on what app renders the page.
+
+## Proposal
+
+We upload the assets to a central location. For example, if both `collections` and `government-frontend` have a file named `foo.css`, they will upload the file to `assets.publishing.service.gov.uk/shared/foo.css`. Because [Rails fingerprints assets](http://guides.rubyonrails.org/asset_pipeline.html#what-is-fingerprinting-and-why-should-i-care-questionmark), the asset actually uploaded will be `assets.publishing.service.gov.uk/shared/foo-8d811b8c3badbc0b0e2f6e25d3660a96cc0cca7993e6f32e98785f205fc40907.css`, unless the files are different, in which case they'll have different fingerprints and won't overwrite.
+
+How to achieve this:
+
+* In govuk-app-deployment, we configure capistrano to precompile the assets and then upload the resulting files to a central location.
+* Proxy assets.publishing.service.gov.uk/shared to this location
+* Point applications to use this URL as asset host so that it's used in production
+
+Upsides
+
+* We remove Nginx config to proxy to individual applications, perhaps removing some hops
+* Whenever there's an asset to share, it will be shared automatically. If an application copies an image from another app, on production they'll have the same URL.
+* Each application ensures that its own assets are present on the shared bucket.
+* Applications don't serve files anymore.
+
+Downsides
+
+* We'll end up with a directory full of assets which are of unknown origin.
```

Comment:
We could (for example) get the script to write a log of assets to a Github repo.  We could then have an icinga check to alert us if anything else was in the bucket unexpectedly (if we thought we needed it)

Reply:
I'm not too worried about this anymore actually. We could easily nuke the bucket and start over by deploying all the applications, so cleaning up would be possible if we ever need it (we don't I think).

Reply:
Could we send the Capistrano logs to kibana? You should be able to search a version hash and find the thing that created it.

Reply:
Yes, good idea!